source/common/config/metadata.h

```diff
@@ -116,11 +116,18 @@ template <typename factoryClass> class TypedMetadataImpl : public TypedMetadata
    */
   void populateFrom(const envoy::config::core::v3::Metadata& metadata) {
     auto& data_by_key = metadata.filter_metadata();
+    auto& typed_data_by_key = metadata.typed_filter_metadata();
     for (const auto& [factory_name, factory] :
          Registry::FactoryRegistry<factoryClass>::factories()) {
-      const auto& meta_iter = data_by_key.find(factory_name);
-      if (meta_iter != data_by_key.end()) {
-        data_[factory->name()] = factory->parse(meta_iter->second);
+      const auto& typed_meta_iter = typed_data_by_key.find(factory_name);
+      // Struct is deprecated in favor of Any.
+      if (typed_meta_iter != typed_data_by_key.end()) {
+        data_[factory->name()] = factory->parse(typed_meta_iter->second);
```

Comment:
I am not sure if this this the right action here: this means for existing typed_metadata factories, all the registered factories will find the factory class and return NUllptr when there is no overrides. System might break without us knowing. 

I suggest we either: 
*make parse(const ProtobufWkt::Any& data) a pure virtual function, so users have to provide an implementation for Any, or 
* fallback to lookup data_by_key if the return value is nullptr?

I think making the parse(const ProtobufWkt::Any& data) a pure virtual function might be more desirable. 
the Parse(const ProtobufWkt::Any& data) can legally return nullptr in which case it's hard to differentiate if it's the default impl working, or the override method returns a "valid" null.

Reply:
I think it will make transitioning easier if the fallback to parsing the Struct happens in both cases, when there is no key in the Any metadata and when the Any `parse` returns nullptr to indicate that it is not implemented (parse throws on parsing errors).

Reply:
fallback to Struct on "nullptr" parsing result works for us. 

But it does have the ambiguity issue that now: if the new "parse method" is implemented and returns nullptr, shall we fallback to Struct anyway? 

Let's comment the assumption here. and in the method description.

Reply:
Thanks for the good comments! Yes, I modified the code to support the behavior as suggested. Please check the latest code.